docs(#128): improve documentation how the plugin works (#147)
mercurius-js/cache: README.md

125k requests in 10.03s, 43.7 MB read
```

## Corner cases resolutions

> It could be possible, especially if we are working with Federation, to come across different return types, for instance, String instead of Int, etc. [Here](https://github.com/mercurius-js/cache/issues/128) is the relative issue.

Reviewer: It could be possible, especially when working with Federation, to encounter varying return types. For instance, one might come across a situation where a String is returned instead of an Int. [Here](https://github.com/mercurius-js/cache/issues/128) is the related issue.

> This situation should be possible only in a development environment before production. But in case we face this situation, there is possible workaround to follow to prevent Mercurius from returning the cached item instead of the real result.

Reviewer: This situation is anticipated to arise primarily during development, before going into production. However, if confronted with this circumstance, there exists a viable workaround that can be employed to prevent Mercurius from returning the cached item instead of the actual result.

> **Important** This solution must be implemented before going in production, when the item is cached the only way to exit from the situation is invalidate the cache or wait the expiration item.

Reviewer: **Important** This solution must be implemented before going into production. Once an item is cached, the only way to rectify the situation is either by invalidating the cache or awaiting the item's expiration.

### Solution

> This solution uses the `onResolution` hook expose by Mercurius. If the response contains errors, the code invalidates the cache, using the `clear` method.

Reviewer: The solution uses the `onResolution` hook provided by Mercurius. If the response contains errors, the code invalidates the cache using the `clear` method.

```js
app.graphql.addHook("onResolution", async (execution, ctx) => {
  if (execution.errors) {
    ctx.reply.log.error("invalidating cache");
    await app.graphql.cache.clear();
  }
});
```

> **N.B.** This solution invalidates all the cache not only the item that contains the error.

Reviewer: **N.B.** This solution invalidates the entire cache, not only the item that contains the error.

125k requests in 10.03s, 43.7 MB read
```

## More info about how this plugin works

> This plugin caches the resolver's results. The plugin has nothing to do with validating the schema, so if you have a resolver that returns a `String` but in your schema it indicates that the resolver returns an `Int`, the plugin will cache the `String`, and when you call the resolver again, the plugin will return the cached `String` instead of the `Int`. 

Reviewer 2: ```suggestion

This plugin caches the result of the  resolver but if the resolver returns a type that is incompatible with the schema return type, the plugin will cache the invalid return value. When you call the resolver again, the plugin will return the cached value, thereby caching the validation error.

This issue may be exacerbated in a federation setup when you don't have full control over the implementation of federated schema and resolvers.
```

> As a result, the query will return a Schema Validation error and not the right result.

Reviewer 2: remove this sentence as I included it in the comment above

Here you can find an example of the problem.
```js
'use strict'

const fastify = require('fastify')
const mercurius = require('mercurius')
const cache = require('mercurius-cache')

const app = fastify({ logger: true })

const schema = `
  type Query {

>     add(x: Int, y: Int): Int  // <-- this is a Int

Author: I created a simple example with a `getNumber` solver.
I cannot use the `hello` resolver because if I set the return type as `String` and then return an `Int`, everything works smoothly because the number converts to a `String`.
I created a `getNumber` to make sense of the example.

    hello: String
  }
`

const resolvers = {
  Query: {
    async add (_, { x, y }, { reply }) {
      reply.log.info('add called')

      return 'not a number' // <-- this is a String
    }
  }
}

app.register(mercurius, {
  schema,
  resolvers
})

> // cache query "add" responses for 10 seconds

Reviewer 2: remove unnecessary comment

app.register(cache, {
  ttl: 10,
  policy: {
    Query: {
      add: true
      // note: it cache "add" but it doesn't cache "hello"
    }
  }
})

> app.listen({ port: 3000, host: '127.0.0.1' })

Reviewer 2: remove host, unnecessary

```

> If you come across this problem, to rectify the data in the cache, follow these instructions:

Reviewer 2: ```suggestion
If you come across this problem, you will first need to fix your code. Then you have two options:
```

> 1. Are you using an **In-Memory** cache? If yes, after the new deployment, the cache will be cleared, and everything will be ok.

Reviewer 2: ```suggestion
1. If you are you using an in-memory cache, it will be cleared at the next start of the application, so the impact of this issue will be limited
```

> 2. Are you using a **Redis** cache? If yes, after the new deployment, you manually invalidate the cache in Redis or wait for the TTL to expire.

Reviewer 2: ```suggestion
2. If you are you using the Redis cache, you will need to manually invalidate the cache in Redis or wait for the TTL to expire
```
